### PR TITLE
refactor: convert helpers function to getters

### PR DESCRIPTION
**Pull request by krskibin**
# 2020-03-11 update
Due to performance and stability problems, we agreed that our priority is to make our codebase as simple as possible and also keep the highest performance in the mobile application. For this reason, we decided that the returned "Helper functions" from `from` would not be based on the passed to itself argument. Instead of this, each composable will have access to a Helper functions prepared especially for them.

Most common usage:
```js
import useProduct from '@vue-storefront/commercetools';
setup() {
  const { helpers, product } = useProduct();
  return {
    name: helpers.getName(product);
  }
}
```
with two diffrent composables:
```js
import useProduct from '@vue-storefront/commercetools';
import useCart from '@vue-storefront/commercetools';

setup() {
  const { productHelpers: helpers, product } = useProduct();
  const { cartHelpers: helpers, product } = useCart();
  
  return {
    name: helpers.getName(product.value);
    cartProducts: cartHelpers.getProducts(cart.value)
  }
}
```

### CODE REVIEW

File: packages/boilerplate/composables/src/getters/_utils.ts
Comment (Qrzy) [inline packages/boilerplate/composables/src/getters/_utils.ts]: This is rather `wrap` than `unwrap` since it's wrapping element with `Ref` if it's not wrapped already.

File: packages/boilerplate/composables/src/getters/cartHelpers.ts
Comment (Qrzy) [inline packages/boilerplate/composables/src/getters/cartHelpers.ts]: Should those files be called `...Helpers`? This doesn't conform to parent directory name, `getters`. 
Would you consider just removing the `Helpers` postfix from filename?

File: packages/boilerplate/composables/src/getters/index.ts
Comment (Qrzy) [inline packages/boilerplate/composables/src/getters/index.ts]: And then also a change here would be needed, `productHelpers` -> `productGetters`. 
What do you think?

File: packages/boilerplate/composables/src/getters/cartHelpers.ts
Comment (filrak) [reply packages/boilerplate/composables/src/getters/cartHelpers.ts]: +1 for getters

File: packages/commercetools/composables/src/useUser/index.ts
Comment (filrak) [inline packages/commercetools/composables/src/useUser/index.ts:55]: talk to Konrad about that

Comment (filrak) [reply packages/commercetools/composables/src/useUser/index.ts:55]: Also the typesc here should come from the factory, same as it is with other composable properties

File: packages/core/theme-module/theme/components/CartSidebar.vue
Comment (filrak) [inline packages/core/theme-module/theme/components/CartSidebar.vue]: i think it will look cleaner if you will return it like this:
```
return {
  productName: cartGetters.getProductName(product)
}
```
- you will see imemdiately what is available in a template
- no need to use ugly `value`

File: packages/core/theme-module/theme/components/CartSidebar.vue
Comment (filrak) [inline packages/core/theme-module/theme/components/CartSidebar.vue]: same, can be in return statemant

File: packages/core/theme-module/theme/components/RelatedProducts.vue
Comment (filrak) [inline packages/core/theme-module/theme/components/RelatedProducts.vue:55]: why its a function? also its not a function so it should be `productSlug`

File: packages/core/theme-module/theme/components/TopBar.vue
Comment (filrak) [inline packages/core/theme-module/theme/components/TopBar.vue:45]: this is the right usage ;)

File: packages/core/theme-module/theme/components/checkout/CartPreview.vue
Comment (filrak) [inline packages/core/theme-module/theme/components/checkout/CartPreview.vue]: all here should be just
`cartProductName: cartGetters.getProductName(product)`

File: packages/core/theme-module/theme/pages/Category.vue
Comment (filrak) [inline packages/core/theme-module/theme/pages/Category.vue]: same

File: packages/core/theme-module/theme/pages/Checkout/Shipping.vue
Comment (filrak) [inline packages/core/theme-module/theme/pages/Checkout/Shipping.vue:176]: no `value` in ui, and that should be in the form suggested previously

File: packages/core/theme-module/theme/pages/Product.vue
Comment (filrak) [inline packages/core/theme-module/theme/pages/Product.vue]: same

**Review by filrak — CHANGES_REQUESTED** [review]
Overally outstanding job! Congrats!

Just keep consistency in how you're exposing getters to the UI

File: packages/boilerplate/composables/src/getters/_utils.ts
Comment (krskibin) [reply packages/boilerplate/composables/src/getters/_utils.ts]: Yes, thats right ;P

File: packages/boilerplate/composables/src/getters/cartHelpers.ts
Comment (krskibin) [reply packages/boilerplate/composables/src/getters/cartHelpers.ts]: Yes, I decided to add helpers at the end, because of the naming conflict with the getters from the composables, but you right, maybe I would change filenames to just `product, cart, ...etc` and in composables to `import {productStatics (productFn)} from './gettes'`. What do you think?

Comment (krskibin) [reply packages/boilerplate/composables/src/getters/cartHelpers.ts]: With @filrak we decided to rename it to `productPlainGetters`

File: packages/boilerplate/composables/src/getters/index.ts
Comment (krskibin) [reply packages/boilerplate/composables/src/getters/index.ts]: I will change it to `productPlainGetters`, as I wrote in comment above.

File: packages/commercetools/composables/src/useUser/index.ts
Comment (krskibin) [reply packages/commercetools/composables/src/useUser/index.ts:55]: We will refactor the useUser after Konards changes will be merged, for now, I think we can keep it as it is.

File: packages/boilerplate/composables/src/getters/cartHelpers.ts
Comment (patzick) [reply packages/boilerplate/composables/src/getters/cartHelpers.ts]: I believe filename should be same as getters:
`productGetters -> geetters/productGetters.ts`
easy to find when you're interested with that particular getter. Just `product` wouldn't be enough and `productPlainGetters` wouldn't be any clue

File: packages/boilerplate/composables/src/getters/cartHelpers.ts
Comment (patzick) [inline packages/boilerplate/composables/src/getters/cartHelpers.ts]: includeAttributes should be in object. Overall
`getSomething(source, options)`
includeAttributes is in this case one of the properties here.
Also should be optional

File: packages/commercetools/composables/src/getters/cartHelpers.ts
Comment (patzick) [inline packages/commercetools/composables/src/getters/cartHelpers.ts]:
```suggestion
import { getAttributes } from './productGetters';
```

File: packages/boilerplate/composables/src/getters/index.ts
Comment (patzick) [inline packages/boilerplate/composables/src/getters/index.ts]: With default export, there is no tree-shaking here. Please change the lines above as an example:
```ts
export * as userHelpers from './userHelpers';
```

File: packages/commercetools/composables/src/useUser/index.ts
Comment (patzick) [reply packages/commercetools/composables/src/useUser/index.ts:55]: this shouldn't be here, just something at return like
```
return {
  // props
  productGetters: makeComputedGetters(productGetters)
}
```
otherwise it's an implementation nightmare :)

File: packages/core/factories/src/useCartFactory.ts
Comment (patzick) [inline packages/core/factories/src/useCartFactory.ts]: we need to allow here to add more helpers

**Review by patzick — CHANGES_REQUESTED** [review]
Not reviewed all yet, leeft few comment and waiting for "to review" state :)
Just please remember: 
- if there is any additional param in getter needed, then should be as an attribute of second `options` param, which is an optional object

File: packages/core/factories/src/useCartFactory.ts
Comment (krskibin) [reply packages/core/factories/src/useCartFactory.ts]: Now, this type is inferred core/interfaces

File: packages/core/interfaces/src/index.ts
Comment (mkucmus) [inline packages/core/interfaces/src/index.ts:185]: Shopware and many other commerce platforms have a CartItem entity instead of Product directly.
Though the product is the most obvious type of cart item, there is something like a coupon, etc. 
We just need to keep in mind that fact. Maybe a little bit deeper level of abstraction is needed here? We are going to face this problem shortly ;)

File: packages/core/interfaces/src/index.ts
Comment (mkucmus) [inline packages/core/interfaces/src/index.ts:170]: maybe including the ProductType (simple, parent, etc) is a good way to recognize which product we are dealing with,
so `getType` method is maybe a solution here.

File: packages/core/interfaces/src/index.ts
Comment (mkucmus) [inline packages/core/interfaces/src/index.ts:210]: what about BillingMethod stuff?

File: packages/core/interfaces/src/index.ts
Comment (mkucmus) [inline packages/core/interfaces/src/index.ts:218]: `getItems` to get the ordered lines of products/items (including things like totals have - for instane, shipping price line)

File: packages/core/interfaces/src/index.ts
Comment (mkucmus) [reply packages/core/interfaces/src/index.ts:170]: the second thing is that many (magento, shopware for sure) ecommerce platforms have related products set on particular Product entity as one to many relation. Maybe we need to have a method like `getRelatedProducts` method as well.

File: packages/core/interfaces/src/index.ts
Comment (mkucmus) [inline packages/core/interfaces/src/index.ts:173]: Unless we have to deal with regular price only, we need to return a object of prices (regular: 100$, finalPrice: 80%), or add some other getter for special price - after discount, for example.

File: packages/core/interfaces/src/index.ts
Comment (mkucmus) [inline packages/core/interfaces/src/index.ts:17]: are we going to not using the error ref in case of problems? (regarding all composables which can do the DAO calls)

**Review by filrak — APPROVED** [review]

File: packages/core/theme-module/theme/components/checkout/CartPreview.vue
Comment (filrak) [reply packages/core/theme-module/theme/components/checkout/CartPreview.vue]: still unresolved

File: packages/core/interfaces/src/index.ts
Comment (krskibin) [reply packages/core/interfaces/src/index.ts:185]: Sure, there is also a CART_ITEM type, which I can use here.

File: packages/core/utils/src/makeComputedGetters/index.ts
Comment (patzick) [inline packages/core/utils/src/makeComputedGetters/index.ts]: the return type is not unknown, it's Readonly<unknown> :)

File: packages/core/utils/__tests__/wrap.spec.ts
Comment (patzick) [inline packages/core/utils/__tests__/wrap.spec.ts]:
```suggestion
describe('[CORE utils] wrap', () => {
```

File: packages/core/utils/__tests__/makeComputedGetters.spec.ts
Comment (patzick) [inline packages/core/utils/__tests__/makeComputedGetters.spec.ts]: shouldn't be read-only computed?

File: packages/core/interfaces/src/index.ts
Comment (patzick) [inline packages/core/interfaces/src/index.ts]: please fix this formatting, there is something wrong here

File: packages/core/interfaces/src/index.ts
Comment (patzick) [reply packages/core/interfaces/src/index.ts:173]: this is important!

File: packages/core/interfaces/src/index.ts
Comment (patzick) [reply packages/core/interfaces/src/index.ts:170]: +1
PRODUCT_TYPES enum + getType sounds great!

File: packages/core/factories/src/useCartFactory.ts
Comment (patzick) [inline packages/core/factories/src/useCartFactory.ts:120]: makeComputedGetters should be used here

File: packages/core/factories/src/useCategoryFactory.ts
Comment (patzick) [inline packages/core/factories/src/useCategoryFactory.ts:29]: makeComputedGetters should be used here

File: packages/core/factories/src/useProductFactory.ts
Comment (patzick) [inline packages/core/factories/src/useProductFactory.ts:43]: makeComputedGetters should be used here

**Review by patzick — CHANGES_REQUESTED** [review]
few notes, yet `packages/core/interfaces/src/index.ts` Getters interfaces need to be discussed with @filrak and @andrzejewsky.

Really great job so far! Love `makeComputedGetters`! :)

File: packages/boilerplate/composables/src/getters/cartHelpers.ts
Comment (krskibin) [reply packages/boilerplate/composables/src/getters/cartHelpers.ts]: Outside the scope of the task

File: packages/core/factories/src/useCartFactory.ts
Comment (krskibin) [reply packages/core/factories/src/useCartFactory.ts]: To discuss in another task

File: packages/core/utils/src/makeComputedGetters/index.ts
Comment (krskibin) [reply packages/core/utils/src/makeComputedGetters/index.ts]: Changed to `ComputedProperty<any>`

File: packages/core/utils/__tests__/wrap.spec.ts
Comment (krskibin) [reply packages/core/utils/__tests__/wrap.spec.ts]: Fixed!

File: packages/core/utils/__tests__/makeComputedGetters.spec.ts
Comment (krskibin) [reply packages/core/utils/__tests__/makeComputedGetters.spec.ts]: For testing proposes, I used ref to check wheater the computed properties returned the reactive value.

File: packages/core/interfaces/src/index.ts
Comment (krskibin) [reply packages/core/interfaces/src/index.ts]: fixed

File: packages/core/factories/src/useProductFactory.ts
Comment (krskibin) [reply packages/core/factories/src/useProductFactory.ts:43]: then the factory would need a type directly from integration, wouldn't it be better to send ready computed getters?

**Review by andrzejewsky — CHANGES_REQUESTED** [review]
first of all, test coverage dropped by -4%, too much, that means you have to ignore some files (if it doesn't make sense to count coverage for them) or just write few lines of tests, thanks!

File: packages/commercetools/composables/src/useProduct/index.ts
Comment (patzick) [inline packages/commercetools/composables/src/useProduct/index.ts:21]: it shouldn't be used here but in factories under the hood 
image

File: packages/core/factories/src/useProductFactory.ts
Comment (patzick) [reply packages/core/factories/src/useProductFactory.ts:43]: the idea of factory is to take care of repeatable logic and not force integrations to do it on their own :)

File: packages/core/utils/src/makeComputedGetters/index.ts
Comment (patzick) [inline packages/core/utils/src/makeComputedGetters/index.ts:5]:
```suggestion
export interface ComputedGetters {
```

**Review by patzick — CHANGES_REQUESTED** [review]
just moving `makeComputedGetters` into factories + change core/interfaces for getters, which was sent on slack

**Comment by andrzejewsky** [comment]
closed here: https://github.com/DivanteLtd/next/pull/297

**Pull request closed without merging**
